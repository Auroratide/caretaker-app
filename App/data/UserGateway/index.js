export class Permissions {
    static FOCUS = 'focus';
    static ADMIN = 'admin';
    static HELPER = 'helper';
}

const createUserGateway = function() {
    let allUsers = null;

    return {
        currentUser: null,

        getAll() {
            if(allUsers === null) {
                //api call goes here
                allUsers = [
                    {name: 'Sarah', identifier: 'sarah@emailprovider.com', permission: Permissions.FOCUS, roles: [], phone:'123-456-7890'},
                    {name: 'Sarah', identifier: 's', permission: Permissions.FOCUS, roles: [], phone:'123-456-7890'},
                    {name: 'Caroline', identifier: 'caroline@woahdude.com', permission: Permissions.ADMIN, roles: ['driver', 'coordinator'], phone:'123-456-7890'},
                    {name: 'Jack', identifier: 'jack@coolwebsite.com', permission: Permissions.HELPER, roles: ['active friend']},
                    {name: 'Austin', identifier: 'austin@yeehaw.com', permission: Permissions.HELPER, roles: [], phone:'123-456-7890'}
                ]
            }
            return allUsers;
        },

        getName(identifier) {
            const user = allUsers.find((user) => user.identifier === identifier)
            return user ? user.name : 'TBD';
        },

        getRoles() {
            return this.currentUser.roles
        },

        signIn(email) {
            //make call to google oauth and store the results
            let results = {
                email: email
            };
            let userFromBackend = this.getAll().find((user) => user.identifier === results.email);
            if(typeof userFromBackend !== 'undefined') {
                this.currentUser = userFromBackend;
                return true;
            }
            return false;
        },

        isSelf(identifier) {
            return this.currentUser.identifier === identifier;
        },

        isAdmin() {
            return this.currentUser.permission === Permissions.ADMIN
                || this.currentUser.permission === Permissions.FOCUS;
        },

        getCurrentPermissions() {
            return this.currentUser.permission;
        },

        inviteUser(email) {
            //api call to invite the user
            alert('invited ' + email + '!');
        }
    }
}

export default createUserGateway();
