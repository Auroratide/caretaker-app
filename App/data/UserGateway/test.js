import UserGateway from './index.js';

describe('UserGateway', () => {

    it('should return false when attempting to sign in with a non-existent user', () => {
        expect(UserGateway.login('TBD')).toBe(false);
    });

    it('should return true when sign in is successful', () => {
        expect(UserGateway.login('sarah@emailprovider.com')).toBe(true);
    });

    it('should assign TBD as a name when TBD is the identifier', () => {
        expect(UserGateway.getName('TBD')).toBe('TBD');
    })
});
