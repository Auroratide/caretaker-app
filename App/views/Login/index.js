import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  Button,
  TextInput
} from 'react-native';
import UserGateway from '../../data/UserGateway';

export default class Login extends Component {

    static navigationOptions = {
        title: 'Login'
    }

    constructor() {
        super();
        this.state = {
            email: null
        }
    }

    _onUpdateEmail = (text) => {
        this.setState({email: text})
    }

    _onLogin = () => {
        if(UserGateway.login(this.state.email)) {
            this.props.navigation.navigate('Home');
        } else {
            alert('Not an approved user.')
        }
    }

    render() {
        return (
            <View style={styles.container}>
                <Text style={styles.title}>Dummy Login Page</Text>
                <View style={styles.inputContainer}>
                    <View style={styles.inputSpacer}></View>
                    <TextInput
                        style={styles.input}
                        onChangeText={this._onUpdateEmail.bind(this)}
                        placeholder='Enter email'/>
                    <View style={styles.inputSpacer}></View>
                </View>
                <Button title="Login" onPress={this._onLogin.bind(this)} />
            </View>
        );
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center'
    },
    title: {
        fontSize: 24,
        textAlign: 'center',
        margin: 15,
    },
    inputContainer: {
        flexDirection: 'row'
    },
    inputSpacer: {
        flex: .2
    },
    input: {
        flex: .6
    }
});
