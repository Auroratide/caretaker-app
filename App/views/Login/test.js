import React, {Component} from 'react';
import { Button } from 'react-native';
import renderer from 'react-test-renderer';
import UserGateway from '../../gateway/user.js';
import Login from './index.js';

describe('Login Screen', () => {

    let navigation;

    beforeEach(() => {
        navigation = {
            navigate: jest.fn()
        }
    });

    it('should set its state to the value of the text box', () => {
        const login = renderer
            .create(<Login navigation={navigation}/>)
            .getInstance();

        login._onUpdateEmail('text');

        expect(login.state.email).toBe('text');
    });

    it('should sign into the user gateway when the button is clicked', () => {
        const spy = jest.spyOn(UserGateway, 'signIn');
        const login = renderer
            .create(<Login navigation={navigation}/>)
            .getInstance();
        login.state.email = 'test@email.com';

        login._onLogin();

        expect(spy).toHaveBeenCalledWith('test@email.com');
    });

    it('should navigate to the home screen when the button is clicked', () => {
        const login = renderer
            .create(<Login navigation={navigation}/>)
            .getInstance();

        login._onLogin();

        expect(navigation.navigate).toHaveBeenCalledWith('Home');
    });

    it('should match the snapshot', () => {
        const login = renderer.create(<Login navigation={navigation} />);

        expect(login).toMatchSnapshot();
    });
})
