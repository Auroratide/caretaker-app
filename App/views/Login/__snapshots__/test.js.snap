// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen should match the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 24,
        "margin": 15,
        "textAlign": "center",
      }
    }
  >
    Dummy Login Page
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 0.2,
        }
      }
    />
    <TextInput
      onChangeText={[Function]}
      placeholder="Enter email"
      style={
        Object {
          "flex": 0.6,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 0.2,
        }
      }
    />
  </View>
  <View
    accessibilityComponentType="button"
    accessibilityLabel={undefined}
    accessibilityTraits={
      Array [
        "button",
      ]
    }
    accessible={true}
    hitSlop={undefined}
    isTVSelectable={true}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#0C42FD",
              "fontSize": 18,
              "padding": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
